### PR TITLE
Add async version of `Process.popen()`

### PR DESCRIPTION
**Pull request by pcbeard**
This provides a new version of the `popen()` convenience method:

```swift
static public func popen(arguments: [String], environment: [String: String] = ProcessEnv.vars,
                             completion: @escaping (Result<ProcessResult, Swift.Error>) -> Void)
```

This allows the caller to asynchronously receive a `ProcessResult` or `Swift.Error`. This was implemented to reduce the number of active threads necessary to execute multiple sub-processes simultaneously. A new unit test is provided that tests the API. In addition, enhancements were made to the `withTemporaryDirectory()` and `withTemporaryFile()` APIs to make them easier to use from asynchronous contexts:

```swift
public func withTemporaryFile<Result>(
  dir: AbsolutePath? = nil, prefix: String = "TemporaryFile", suffix: String = "", _ body: (TemporaryFile, @escaping (TemporaryFile) -> Void) throws -> Result
) throws -> Result

public func withTemporaryDirectory<Result>(
    dir: AbsolutePath? = nil, prefix: String = "TemporaryDirectory" , _ body: (AbsolutePath, @escaping (AbsolutePath) -> Void) throws -> Result
) throws -> Result
```

Both of these new functions pass the callback block a `@escaping` reference to a cleanup block, to allow the cleanup to be deferred until a later time. The old API uses a Bool argument to disable cleanup, and makes it the caller's responsibility to delete the temporary resources. The cleanup block reduces code duplication.

### CODE REVIEW

**Comment by pcbeard** [comment]
This PR depends on changes in https://github.com/apple/swift-corelibs-foundation/pull/3113 to function correctly on the Windows platform, which uses `Foundation.Process` under the hood.

File: Sources/TSCBasic/Process.swift [hunk 1]
```diff
@@ -677,24 +688,40 @@ public final class Process {
...
+        case .failure(let error):
+            throw error
+        case .success(let result):
+            return result
+        }
+    }
+
+    /// Executes the process I/O state machine, calling completion block when finished.
+    private func waitUntilExit(_ completion: @escaping (Result<ProcessResult, Swift.Error>) -> Void) {
         self.stateLock.lock()
```
Comment (pcbeard) [inline Sources/TSCBasic/Process.swift:709]: This could probably use a better name.

File: Sources/TSCBasic/Process.swift [hunk 2]
```diff
@@ -180,10 +180,10 @@ public final class Process {
     // process execution mutable state
     private enum State {
         case idle
-        case readingOutputThread(stdout: Thread, stderr: Thread?)
         case readingOutputPipe(sync: DispatchGroup)
         case outputReady(stdout: Result<[UInt8], Swift.Error>, stderr: Result<[UInt8], Swift.Error>)
         case complete(ProcessResult)
+        case failed(Swift.Error)
```
Comment (tomerd) [inline Sources/TSCBasic/Process.swift:186]: is there a different Error type in this code base that requires qualifying it as Swift.Error?

File: Sources/TSCBasic/Process.swift [hunk 3]
```diff
@@ -677,24 +688,40 @@ public final class Process {
...
+        switch processResult! {
+        case .failure(let error):
+            throw error
+        case .success(let result):
+            return result
+        }
+    }
+
+    /// Executes the process I/O state machine, calling completion block when finished.
+    private func waitUntilExit(_ completion: @escaping (Result<ProcessResult, Swift.Error>) -> Void) {
```
Comment (tomerd) [inline Sources/TSCBasic/Process.swift:708]: should we allow the caller to specify a queue instead of hard coding one? the caller may be some preferences

File: Sources/TSCBasic/Process.swift [hunk 4]
```diff
@@ -677,24 +688,40 @@ public final class Process {
...
+            processResult = result
+            group.leave()
+        }
+        group.wait()
+        switch processResult! {
+        case .failure(let error):
+            throw error
+        case .success(let result):
+            return result
+        }
```
Comment (tomerd) [inline Sources/TSCBasic/Process.swift:704]: can be simplified to `return try processResult.get()`

File: Sources/TSCBasic/Process.swift [hunk 5]
```diff
@@ -230,6 +230,7 @@ public final class Process {
     // process execution mutable state
     private var state: State = .idle
     private let stateLock = Lock()
+    private static let stateQueue = DispatchQueue(label: "ProcessStateQueue")
```
Comment (tomerd) [inline Sources/TSCBasic/Process.swift:233]: we should consider allowing the user to specify the queue (with a default). also this is a serial one, correct?

File: Sources/TSCBasic/Process.swift [hunk 6]
```diff
@@ -677,24 +688,40 @@ public final class Process {
...
-            return result
-        case .readingOutputThread(let stdoutThread, let stderrThread):
-            self.stateLock.unlock() // unlock early since output read thread need to change state
-            // If we're reading output, make sure that is finished.
-            stdoutThread.join()
-            stderrThread?.join()
-            return try self.waitUntilExit()
+            completion(.success(result))
+        case .failed(let error):
+            defer { self.stateLock.unlock() }
```
Comment (tomerd) [inline Sources/TSCBasic/Process.swift:718]: why defer and not unlock directly?

File: Sources/TSCBasic/Process.swift [hunk 7]
```diff
@@ -677,24 +688,40 @@ public final class Process {
...
+            completion(.failure(error))
         case .readingOutputPipe(let sync):
-            self.stateLock.unlock() // unlock early since output read thread need to change state
-            sync.wait()
-            return try self.waitUntilExit()
+            defer { self.stateLock.unlock() }
```
Comment (tomerd) [inline Sources/TSCBasic/Process.swift:721]: why defer and not unlock directly?

File: Sources/TSCBasic/Process.swift [hunk 8]
```diff
@@ -648,22 +655,26 @@ public final class Process {
...
             } else {
                 pendingLock.withLock {
                     pending = .success([])  // no stderr in this case
                 }
             }
+            
             // first set state then start reading threads
             self.stateLock.withLock {
-                self.state = .readingOutputThread(stdout: stdoutThread, stderr: stderrThread)
+                self.state = .readingOutputPipe(sync: group)
```
Comment (tomerd) [inline Sources/TSCBasic/Process.swift:675]: we should probably rename this to something else than readingOutputPipe if reusing it for the thread bases scenario as well, how about `readingOutput`

File: Sources/TSCBasic/TemporaryFile.swift [hunk 9]
```diff
@@ -213,14 +237,14 @@ extension MakeDirectoryError: CustomNSError {
...
-///             If `body` has a return value, that value is also used as the
-///             return value for the `withTemporaryDirectory` function.
+///           If `body` has a return value, that value is also used as the
+///           return value for the `withTemporaryDirectory` function.
+///           The cleanup block should be called when the temporary directory is no longer needed.
 ///
 /// - Throws: MakeDirectoryError and rethrows all errors from `body`.
 public func withTemporaryDirectory<Result>(
-    dir: AbsolutePath? = nil, prefix: String = "TemporaryDirectory", removeTreeOnDeinit: Bool = false , _ body: (AbsolutePath) throws -> Result
+    dir: AbsolutePath? = nil, prefix: String = "TemporaryDirectory" , _ body: (AbsolutePath, @escaping (AbsolutePath) -> Void) throws -> Result
```
Comment (tomerd) [inline Sources/TSCBasic/TemporaryFile.swift:247]: can these changes move to a separate PR? or are they related to the async process?

**Review by tomerd — CHANGES_REQUESTED** [review]
thanks @pcbeard this is something I wanted to do for a very long time <3

some feedback inline

File: Sources/TSCBasic/Process.swift (hunk 2 above)
Comment (pcbeard) [reply Sources/TSCBasic/Process.swift:186]: Yes, `ProcessResult.Error`.

File: Sources/TSCBasic/Process.swift (hunk 3 above)
Comment (pcbeard) [reply Sources/TSCBasic/Process.swift:708]: Possibly. Before we were using a .global queue to wake up the dispatch group. Using a shared serial queue seems better. But a shared specified queue might be fine too.

File: Sources/TSCBasic/Process.swift (hunk 4 above)
Comment (pcbeard) [reply Sources/TSCBasic/Process.swift:704]: Thanks for the tips. I'm still learning this pattern.

File: Sources/TSCBasic/Process.swift (hunk 5 above)
Comment (pcbeard) [reply Sources/TSCBasic/Process.swift:233]: Global/concurrent queues are in general discouraged as a dispatch anti-pattern. They can create extra threads unnecessarily. I needed a queue to field the .notify() calls on. It's only a callback when the state machine calls the completion block. We could even do away with `stateLock` possibly.

File: Sources/TSCBasic/Process.swift (hunk 8 above)
Comment (pcbeard) [reply Sources/TSCBasic/Process.swift:675]: Sounds good.

File: Sources/TSCBasic/TemporaryFile.swift (hunk 9 above)
Comment (pcbeard) [reply Sources/TSCBasic/TemporaryFile.swift:247]: They are needed upstream in SPM.

File: Sources/TSCBasic/Process.swift (hunk 6 above)
Comment (pcbeard) [reply Sources/TSCBasic/Process.swift]: Fixed in e5f2e14.

File: Sources/TSCBasic/Process.swift (hunk 7 above)
Comment (pcbeard) [reply Sources/TSCBasic/Process.swift]: Fixed in e5f2e14.

File: Sources/TSCBasic/Process.swift (hunk 5 above)
Comment (pcbeard) [reply Sources/TSCBasic/Process.swift]: Added optional queue parameter in d67a7f0.

File: Sources/TSCBasic/Process.swift (hunk 3 above)
Comment (pcbeard) [reply Sources/TSCBasic/Process.swift:709]: Added optional queue parameter in d67a7f0.

File: Sources/TSCBasic/TemporaryFile.swift (hunk 9 above)
Comment (tomerd) [reply Sources/TSCBasic/TemporaryFile.swift:247]: I am a bit iffy about the `@escaping (AbsolutePath) -> Void` cleanup argument passed into the closure. feel like the boolean cleanup is good enough for the public API and this should an internal implementation details?

Comment (pcbeard) [reply Sources/TSCBasic/TemporaryFile.swift:247]: My rationale for this is to avoid code duplication.

Comment (tomerd) [reply Sources/TSCBasic/TemporaryFile.swift:247]: @pcbeard right. not a big deal, mostly wondering if the variant with the cleanup closure needs to be public, or if the boolean variant is good enough as the public api. i think SwiftPM can just use the boolean variant?

**Review by tomerd — APPROVED** [review]

**Comment by tomerd** [comment]
@swift-ci test

**Comment by pcbeard** [comment]
Please hold off merging this until https://github.com/apple/swift-corelibs-foundation/pull/3113 is merged.

File: Sources/TSCBasic/Process.swift [hunk 10]
```diff
@@ -230,6 +230,8 @@ public final class Process {
     // process execution mutable state
     private var state: State = .idle
     private let stateLock = Lock()
+    private static let sharedCompletionQueue = DispatchQueue(label: "Process.sharedCompletionQueue")
```
Comment (compnerd) [inline Sources/TSCBasic/Process.swift:233]: Would be nice to use `org.swift.tools-support-core.process-completion` as the name for the queue.

File: Sources/TSCBasic/Process.swift [hunk 11]
```diff
@@ -609,8 +613,9 @@ public final class Process {
 
             // Close the local write end of the output pipe.
             try close(fd: outputPipe[1])
-
+
```
Comment (compnerd) [inline Sources/TSCBasic/Process.swift:616]: Bleeding whitespace

File: Sources/TSCBasic/Process.swift [hunk 12]
```diff
@@ -622,11 +627,13 @@ public final class Process {
                             pending = readResult
                         }
                     }
-                } else if let stderrResult = (pendingLock.withLock { pending }) {
+                    group.leave()
+               } else if let stderrResult = (pendingLock.withLock { pending }) {
```
Comment (compnerd) [inline Sources/TSCBasic/Process.swift:631]: Indentation seems off?

File: Tests/TSCBasicTests/ProcessTests.swift [hunk 13]
```diff
@@ -42,20 +42,57 @@ class ProcessTests: XCTestCase {
     }
 
     func testPopen() throws {
+        #if os(Windows)
+        let echo = "echo.exe"
```
Comment (compnerd) [inline Tests/TSCBasicTests/ProcessTests.swift:46]: This worries me - there is no `echo.exe` technically.  It is a MSys/2 binary that is part of git and if you add Git to your path, it might show up or it might now.  On the other hand, the vast majority of the t-s-c tests will just abort anyways, so whats a broken test?

The equivalent of echo is echo, which is a shell builtin.

File: Tests/TSCBasicTests/ProcessTests.swift [hunk 14]
```diff
@@ -42,20 +42,57 @@ class ProcessTests: XCTestCase {
...
 
         // Test buffer larger than that allocated.
         try withTemporaryFile { file in
             let count = 10_000
             let stream = BufferedOutputByteStream()
             stream <<< Format.asRepeating(string: "a", count: count)
             try localFileSystem.writeFileContents(file.path, bytes: stream.bytes)
-            let outputCount = try Process.popen(args: "cat", file.path.pathString).utf8Output().count
+            #if os(Windows)
+            let cat = "cat.exe"
```
Comment (compnerd) [inline Tests/TSCBasicTests/ProcessTests.swift:60]: Similar to echo, `cat.exe` is not a thing.  The equivalent to `cat` is `type` which is a shell builtin.

File: Tests/TSCBasicTests/ProcessTests.swift [hunk 15]
```diff
@@ -42,20 +42,57 @@ class ProcessTests: XCTestCase {
...
+            #endif
+            let outputCount = try Process.popen(args: cat, file.path.pathString).utf8Output().count
             XCTAssert(outputCount == count)
         }
     }
 
+    func testPopenAsync() throws {
+        #if os(Windows)
+        let args = ["where.exe", "where"]
+        let answer = "C:\\Windows\\System32\\where.exe"
```
Comment (compnerd) [inline Tests/TSCBasicTests/ProcessTests.swift:72]: There is no guarantee that the path is `C:\Windows\System32\where.exe` . The correct answer is `%SystemRoot%\System32\where.exe`.  You can use `GetSystemDirectoryW` to get the value for `%SystemRoot%` rather than expanding it via an environment variable.

File: Sources/TSCBasic/Process.swift (hunk 12 above)
Comment (pcbeard) [reply Sources/TSCBasic/Process.swift]: Fixed in f30c446.

File: Sources/TSCBasic/Process.swift (hunk 10 above)
Comment (pcbeard) [reply Sources/TSCBasic/Process.swift]: Fixed in 95e286b.

**Comment by compnerd** [comment]
@swift-ci please smoke test

**Comment by compnerd** [comment]
@swift-ci please test
